src/App.tsx

```diff
@@ -1,26 +1,237 @@
+  }, [todos]);
+
+  useEffect(() => {
+    setLoading(true);
+    todoService
+      .getTodos()
+
+      .then(todoFromServer => setTodos(todoFromServer))
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      })
+      .finally(() => {
+        setLoading(false);
+      });
+  }, []);
+
+  if (!todoService.USER_ID) {
     return <UserWarning />;
   }
 
+  const filteredTodos = filterTodos(todos, filter);
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!title.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setDisable(true);
+    let requestSuccessful = false;
+    const temptodo: Todo = {
+      userId: todoService.USER_ID,
+      title: title.trim(),
+      completed: false,
+      id: 0,
+    };
+
+    setLoadingTodoIds(prev => [...prev, temptodo.id]);
+    setTempTodo(temptodo);
+    todoService
+      .createTodo(temptodo)
+      .then(createdTodo => {
+        setTodos(currentTodos => [...currentTodos, createdTodo]);
+        requestSuccessful = true;
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+
+        setTitle(title);
+      })
+      .finally(() => {
+        setDisable(false);
+        setLoadingTodoIds(prev => prev.filter(id => id !== temptodo.id));
+        setTempTodo(null);
+        if (requestSuccessful) {
+          setTitle('');
+        }
+      });
+  };
+
+  const handleComplitedToDo = (updatedTodo: Todo) => {
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const { id, title, completed } = updatedTodo;
+
+    setLoadingTodoIds(prev => [...prev, id]);
+
+    todoService
+      .upDateTodo(id, { title, completed: !completed })
+      .then(() => {
+        setTodos(
+          todos.map(todo =>
+            todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+          ),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        setLoadingTodoIds(prev => prev.filter(todoId => todoId !== id));
+      });
+  };
+
+  const deleteTodo = (id: number) => {
+    setLoadingTodoIds(prev => [...prev, id]);
+    todoService
+      .deleteTodo(id)
+      .then(() => {
+        setTodos(todos.filter(todo => todo.id !== id));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        setLoadingTodoIds(prev => prev.filter(todoId => todoId !== id));
+      });
+  };
+
+  const clearCompleted = () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    Promise.all(
+      completedTodos.map(todo => {
+        setLoadingTodoIds(prev => [...prev, todo.id]);
+
+        return todoService
+          .deleteTodo(todo.id)
+          .then(() => {
+            setTodos(currentTodos =>
+              currentTodos.filter(t => t.id !== todo.id),
```

Comment:
```suggestion
              currentTodos.filter(todo => t.id !== todo.id),
```

Reply:
I use the todo name to iterate over completedTodos, and when filtering, I compare todo.id with the previous id. That's why I can't use the same name...

Reply:
okay take currentTodo